### PR TITLE
server:  schedule option persistence

### PR DESCRIPTION
**Pull request by nolouch**
PTAL @siddontang @disksing @overvenus

### CODE REVIEW

File: server/leader.go [hunk 1]
```diff
@@ -181,6 +181,13 @@ func (s *Server) campaignLeader() error {
 	defer s.enableLeader(false)
 
 	// Try to create raft cluster.
+	isExist, err := s.kv.loadScheduleOption(s.scheduleOpt)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !isExist {
+		s.kv.saveScheduleOption(s.scheduleOpt)
```
Comment (siddontang) [inline server/leader.go]: check error

File: server/api/config_test.go [hunk 2]
```diff
@@ -89,3 +89,35 @@ func (s *testConfigSuite) TestConfigSchedule(c *C) {
...
+
+func (s *testConfigSuite) TestConfigReplication(c *C) {
+	numbers := []int{1, 3}
+	for _, num := range numbers {
+		cfgs, _, clean := mustNewCluster(c, num)
+		defer clean()
+
+		parts := []string{cfgs[rand.Intn(len(cfgs))].ClientUrls, apiPrefix, "/api/v1/config/replicate"}
+		addr := mustUnixAddrToHTTPAddr(c, strings.Join(parts, ""))
+		resp, err := s.hc.Get(addr)
```
Comment (siddontang) [inline server/api/config_test.go]: can we use readJSON instead?

File: server/api/config_test.go [hunk 3]
```diff
@@ -89,3 +89,35 @@ func (s *testConfigSuite) TestConfigSchedule(c *C) {
...
+		c.Assert(err, IsNil)
+
+		rc.MaxReplicas = 5
+		postData, err := json.Marshal(rc)
+		postURL := []string{cfgs[rand.Intn(len(cfgs))].ClientUrls, apiPrefix, "/api/v1/config/replicate"}
+		postAddr := mustUnixAddrToHTTPAddr(c, strings.Join(postURL, ""))
+		resp, err = s.hc.Post(postAddr, "application/json", bytes.NewBuffer(postData))
+		c.Assert(err, IsNil)
+
+		resp, err = s.hc.Get(addr)
```
Comment (siddontang) [inline server/api/config_test.go]: ditto

File: server/api/config_test.go [hunk 4]
```diff
@@ -89,3 +89,35 @@ func (s *testConfigSuite) TestConfigSchedule(c *C) {
...
+		err = json.Unmarshal(buf, rc)
+		c.Assert(err, IsNil)
+
+		rc.MaxReplicas = 5
+		postData, err := json.Marshal(rc)
+		postURL := []string{cfgs[rand.Intn(len(cfgs))].ClientUrls, apiPrefix, "/api/v1/config/replicate"}
+		postAddr := mustUnixAddrToHTTPAddr(c, strings.Join(postURL, ""))
+		resp, err = s.hc.Post(postAddr, "application/json", bytes.NewBuffer(postData))
+		c.Assert(err, IsNil)
+
```
Comment (siddontang) [inline server/api/config_test.go]: no need close the resp body here?

File: server/leader.go [hunk 5]
```diff
@@ -175,6 +175,16 @@ func (s *Server) campaignLeader() error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	isExist, err := s.kv.loadScheduleOption(s.scheduleOpt)
```
Comment (siddontang) [inline server/leader.go]: extract a function. 
Here we should also handle replication option.

Comment (nolouch) [inline server/leader.go]: scheduleOpt already contains scheduleConf and replicationConf

Comment (siddontang) [inline server/leader.go]: so extract a  function here.

File: server/api/config_test.go [hunk 6]
```diff
@@ -68,24 +68,54 @@ func (s *testConfigSuite) TestConfigSchedule(c *C) {
...
 		sc := &server.ScheduleConfig{}
-		err = json.Unmarshal(buf, sc)
-		c.Assert(err, IsNil)
+		readJSON(resp.Body, sc)
 
 		sc.MaxStoreDownTime.Duration = time.Second
 		postData, err := json.Marshal(sc)
 		postURL := []string{cfgs[rand.Intn(len(cfgs))].ClientUrls, apiPrefix, "/api/v1/config"}
 		postAddr := mustUnixAddrToHTTPAddr(c, strings.Join(postURL, ""))
 		resp, err = s.hc.Post(postAddr, "application/json", bytes.NewBuffer(postData))
```
Comment (siddontang) [inline server/api/config_test.go]: extract a function to handle Post, read all body, and close body.

File: server/leader.go [hunk 7]
```diff
@@ -260,3 +263,17 @@ func (s *Server) resignLeader() error {
 func (s *Server) leaderCmp() clientv3.Cmp {
 	return clientv3.Compare(clientv3.Value(s.getLeaderPath()), "=", s.leaderValue)
 }
+
+func (s *Server) reloadScheduleOption() error {
+	isExist, err := s.kv.loadScheduleOption(s.scheduleOpt)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !isExist {
```
Comment (siddontang) [inline server/leader.go]:
```
if isExist {
    return nil 
}

return s.kv.save
```

**Comment by disksing** [comment]
Seems it has compatibility issues. For instance, if we add a new config value "foobar" with default value 100, it will be 0 in all old clusters.
cc @siddontang

**Comment by siddontang** [comment]
Yes, so we can only allow using this feature after rolling update all PDs.

**Comment by nolouch** [comment]
if we add a schedule option, we need to delete the option key/value firstly and then rolling update.

**Comment by disksing** [comment]
How about do it like this:
When pd-server starts, it loads config from file. And when it's ready to serve as leader, we can merge the config with old config stored in etcd. It can be done easily using json unmarshal. Here's a simple demo: https://play.golang.org/p/X9ohwSQR81

**Comment by siddontang** [comment]
> Seems it has compatibility issues

@disksing I think it doesn't matter here, we still had the same problem before, leader updated configuration, but new leader still used the old one. so we can upgrade it safely.

**Comment by siddontang** [comment]
> if we adjust a schedule option, we need to delete the config key/value firstly and then rolling update

@nolouch why do we need to delete the old config?

**Comment by nolouch** [comment]
PTAL @disksing @siddontang

**Comment by disksing** [comment]
LGTM

File: server/kv.go [hunk 8]
```diff
@@ -86,6 +89,52 @@ func (kv *kv) saveRegion(region *metapb.Region) error {
...
+	cfg.Schedule = *opt.load()
+	cfg.Replication = *opt.rep.load()
+	isExist, err := kv.loadConfig(cfg)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if !isExist {
+		return false, nil
+	}
+	opt.store(&cfg.Schedule)
```
Comment (siddontang) [inline server/kv.go]: any test to cover this?

Comment (nolouch) [inline server/kv.go]: I have added test.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**